### ISSUE TITLE
[MNG-6089] Define a command line option to define settings-security.xml as well as settings.xml as well

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7652]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** opened **[MNG-6089](https://issues.apache.org/jira/browse/MNG-6089?redirect=false)** and commented

It would be nice improvement to add a command line to support the defintion of the `settings-security.xml` like the `settings.xml` which would also help support in Jenkins and of course in many other ways.

---

**Affects:** 3.3.9

**Issue Links:**
- [MNG-5438](https://issues.apache.org/jira/browse/MNG-5438) cli parameter to use a custom path settings-security.xml

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I would expect this file to always be next to the user-settings.xml, so I have my doubts about this option.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

On a user machine I would agree but if you are on a CI solution it would make sense to be able to define different locations...and to offer different `settiings-security.xml` location for different `settings.xml` setups. For example with Config File provider in Jenkins...otherwise you need to have the `settings-security.xml` on each system where you run maven (multiple slaves etc.). which you can already prevent by using the config file provider with the `settings.xml` also with `toolchains.xml`....but not for the `settings-security.xml`.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

We both know several Cloudbees experts which are around here for a long time. I would have expected that such an important issue would already be registered. Unless there's no issue... I wouldn't be surprised is something like this can be fixed in the Config File Provider. In concept I think settings-security are personal and shouldn't be specified at a global level.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Hm...I'm not sure about that..If you take a look into the docs of the config file provider plugin you will find this:

> If your passwords are encrypted by maven, you can store the encrypted passwords in the same configuration - but you have to make sure the settings-security.xml is available to the job, unfortunately maven does not yet provide a way to pass this file aswell, so this file has to be located at the original location required by maven (<user.home>/.m2)

which shows me there is a requirement for such option furthermore I have currently exactly this problem..In particular if you would like to assemble a slave from the master...without manually copying such file on the different machines...Not to think about things like Docker...

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

From my PoV we should do one of three things:

1. The `settings-security.xml` file should always be a sibling of the `settings.xml` file.
2. The `settings-security.xml` file should be checked first as a sibling of the `settings.xml` file and if not present then falling back to `~/.m2/`
3. Provide an option / mechanism to control the location

But actually when we look into the code a bit more carefully...

For using maven in embedded mode we cannot even assume there is a settings.xml file in the first place.

When the maven-release-plugin comes along to capture the effective settings, it therefore cannot be assured of a location of the settings.xml file and has to put the "effective" settings into a temporary location in order to ensure that the build runs with the correct settings.

So we have to take the release plugin concerns into account. It would be bad if the effective settings written to a temporary disk file by the release plugin contained decrypted passwords (IIRC it currently writes the passwords with the `settings-security.xml` encryption into that temporary file)... what would be much better is if the secrets were encrypted using a one-time key that is also thrown away.

We may have to dig into the embedding API to ensure that we do not make assumptions here that invalidate the embedder use as we can have a settings.xml that is never on disk or even an XML file through embedder... so it seems to be that an embedder user may also have a security settings that isn't even an XML file or on disk.

TL;DR

The ability to control the storage location of the master key is a requirement, but the actual mechanism as to how to implement and expose this requires an awareness of how settings are exposed through the embedder API
